### PR TITLE
Update to MXNet 0.12

### PR DESCRIPTION
**Pull request by fhieber**
WIP.
Checks will likely fail until MXNet 0.12.0 is released.
Created a new major version to update some flags. Please let me know if you want to integrate any backwards incompatible changes into this.

### CODE REVIEW

File: sockeye/loss.py [hunk 1]
```diff
@@ -168,87 +163,3 @@ def update(self, labels, preds):
...
-
-        cross_entropy = mx.sym.MakeLoss(cross_entropy, name=C.SMOOTHED_CROSS_ENTROPY)
-        probs = mx.sym.BlockGrad(probs, name=C.SOFTMAX_NAME)
-        return [cross_entropy, probs]
-
-    def create_metric(self) -> "SmoothedCrossEntropyMetric":
-        return SmoothedCrossEntropyMetric(self.loss_config)
-
-
-class SmoothedCrossEntropyMetric(EvalMetric):
```
Comment (mjdenkowski) [inline sockeye/loss.py]: It's great that we can merge smoothed and non-smoothed loss!  We should make sure that `CrossEntropyMetric` can also apply smoothing to match `CrossEntropyLoss`.

Comment (fhieber) [reply sockeye/loss.py]: I thought about that, but its also actually not too bad to have the same loss values for smoothed and non-smoothed gradients for easy comparison. I will look into it.

Comment (mjdenkowski) [reply sockeye/loss.py]: The main place we're using the loss-replicating metrics is for optimizers like Eve that want the exact loss from training.  The idea is that `Loss` classes can return corresponding `Metric` versions that use the same config and return the same results.  For checkpoints, we're still using the vanilla MXNet metrics.

Comment (fhieber) [reply sockeye/loss.py]: Ah you are right, I will put that back in then!

Comment (fhieber) [reply sockeye/loss.py]: I added back the smoothed cross-entropy for the Loss metric. However, this is super slow as we are not caching the label distributions as before in the symbol graph. This probably removes all the benefit we have from the fusioning of the smoothed CE grad in MXNet 0.12.
I will think about a different way for this.

Comment (mjdenkowski) [reply sockeye/loss.py]: Do we have a way to return the loss?  The custom metrics were originally only used for the Eve optimizer, which just needs the loss.

Comment (fhieber) [reply sockeye/loss.py]: The problem is the smoothed loss in general. Explicitly creating the soft labels / label distribution is super expensive. As it is right now, we reduce speed by 60%.

**Comment by tdomhan** [comment]
I remembered what other breaking change I was thinking about adding: Adding another digit or two to the parameter file names to reduce the chance of overflowing at some point. What do you think?

**Comment by fhieber** [comment]
fine with me :) will add that

**Comment by philschulz** [comment]
Seem like mxnet 0.12 has been released (at least it's on pypi). Assuming that I have mxnet 0.12 installed, is it safe to rebase my code on this branch?

**Comment by fhieber** [comment]
I'd say yes you can. I will rebase this PR on latest master.

**Comment by philschulz** [comment]
Great, thanks a lot! I'll rebase tomorrow (it's 9pm in Australia right now).

File: README.md [hunk 2]
```diff
@@ -50,19 +50,13 @@ remaining instructions to work you will need to use `python3` instead of `python
...
+Depending on your version of CUDA, you can do this by running the following:
 ```bash
-> wget https://raw.githubusercontent.com/awslabs/sockeye/master/requirements.gpu-cu75.txt
-> pip install sockeye --no-deps -r requirements.gpu-cu75.txt
-> rm requirements.gpu-cu75.txt
+> wget https://raw.githubusercontent.com/awslabs/sockeye/master/requirements.gpu-cu${CUDA_VERSION}.txt
+> pip install sockeye --no-deps -r requirements.gpu-cu${CUDA_VERSION}.txt
+> rm requirements.gpu-cu${CUDA_VERSION}.txt
 ```
+where `${CUDA_VERSION}` can be `75`, `80`, or `90`.
```
Comment (tdomhan) [inline README.md]: maybe mention the fact that 7.5 = 75, 8.0=80 and 9.0=90

File: sockeye/layers.py [hunk 3]
```diff
@@ -108,6 +112,64 @@ def __call__(self) -> mx.sym.Symbol:
...
+                 num_hidden: int,
+                 num_embed: int,
+                 vocab_size: int,
+                 weight_tying: bool,
+                 embed_weight: Optional[mx.sym.Symbol],
+                 weight_normalization: bool,
+                 prefix: str = '') -> None:
+        self.vocab_size = vocab_size
+
+        self.w = mx.sym.Variable("%scls_weight" % prefix, shape=(vocab_size, num_hidden))
```
Comment (tdomhan) [inline sockeye/layers.py]: how about moving these into the else branch of `weight_tying` to only create it if weight tying is turned off?

Comment (fhieber) [reply sockeye/layers.py]: good call, will change!

File: README.md (hunk 2 above)
Comment (fhieber) [reply README.md]: done!

File: sockeye/loss.py (hunk 1 above)
Comment (fhieber) [reply sockeye/loss.py]: I will leave as is right now as this is correct. If we see that eve + label smoothing is too slow for us we may take another look at this.
